servers/republik/modules/crowdfundings/lib/scheduler/feedback.js

```diff
@@ -0,0 +1,110 @@
+const debug = require('debug')('crowdfundings:lib:scheduler:feedback')
+const Promise = require('bluebird')
+
+const { transformUser } = require('@orbiting/backend-modules-auth')
+const { sendMailTemplate } = require('@orbiting/backend-modules-mail')
+
+const findRecipients = (context) => {
+  return context.pgdb.query(`
+    WITH "eligables" AS (
+      SELECT
+        u.id "userId",
+      
+        -- All periods expressed in days
+        EXTRACT(DAYS FROM SUM(mp."endDate" - mp."beginDate")) "days",
+      
+        -- Days behind (a)
+        EXTRACT(DAYS FROM SUM(LEAST(now(), mp."endDate") - LEAST(now(), mp."beginDate"))) "daysBehind",
+        -- Days ahead (b)
+        EXTRACT(DAYS FROM SUM(GREATEST(now(), mp."endDate") - GREATEST(now(), mp."beginDate"))) "daysAhead",
+      
+        -- is overdue?
+        MAX(mp."endDate") < now() "isOverdue"
+      
+      FROM "users" u
+      -- users memberships
+      JOIN "memberships" m ON m."userId" = u.id
+      -- memberships periods
+      JOIN "membershipPeriods" mp ON mp."membershipId" = m.id
+      
+      -- A membership which is …
+      JOIN "memberships" ma ON ma."userId" = u.id
+        -- … still active
+        AND ma.active = TRUE
+        -- … not cancelled
+        AND ma.renew = TRUE
+      
+      GROUP BY u.id
+    )
+    
+    SELECT *
+    FROM "eligables" e
+    JOIN "users" u ON u.id = e."userId"
+
+    -- Scope to days behind a user between 21 and 27 days (3rd week)
+    WHERE e."daysBehind" BETWEEN 21 AND 27
+  `)
```

Comment:
This query fails for short periods in the past. Let's say for example a user had a (monthly) abo which started "2020-01-02 12:00:00" and ended "2020-01-26 12:00:00" (due to cancelling it at that day immediately). The query above would match it with days behind 24 even though the membership endend a long time ago.

Reply:
This is hindered a little earlier, basically addressing scenarios (a) in which a membership is no longer active but falls into 21-27 day period or (b) a membership which has been cancelled.

```sql
      -- A membership which is …
      JOIN "memberships" ma ON ma."userId" = u.id
        -- … still active
        AND ma.active = TRUE
        -- … not cancelled
        AND ma.renew = TRUE
```